### PR TITLE
feat: delete csv-source

### PR DESCRIPTION
**Pull request by lucafurrer**
Delete a CSV-Source
Does:
- Remove file from S3
- Delete graph of csv.source

Does not (yet): 
- Delete associated tables or prevent deletion when tables are present 

Part of #88

### CODE REVIEW

File: apis/core/lib/ResourceStore.ts [hunk 1]
```diff
@@ -121,4 +140,11 @@ export default class implements ResourceStore {
 
     return member
   }
+
+  delete(id: NamedNode): void {
+    if (!this.__resources.has(id)) {
+      throw new Error('Resource does not exist')
+    }
+    this.__deletedGraphs.add(id)
```
Comment (lucafurrer) [inline apis/core/lib/ResourceStore.ts:148]: Should we remove all nodes of this graph from the store at this moment?

File: apis/core/lib/domain/csv-source/delete.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+export async function deleteSource({
+  resource,
+  store = resourceStore(),
+}: DeleteSourceCommand): Promise<void> {
+  const csvSource = await store.get(resource)
+
+  // TODO: Delete Tables or prevent it
+  // Find related tables
+  // Delete them
```
Comment (lucafurrer) [inline apis/core/lib/domain/csv-source/delete.ts:20]: I suggest to wait until the table implementation is done, to continue on this

File: apis/core/test/ResourceStore.test.ts [hunk 3]
```diff
@@ -165,4 +165,33 @@ describe('ResourceStore', () => {
...
+      const resourceStore = new ResourceStore(client)
+      resourceStore.create(ex.Foo)
+      await resourceStore.save()
+      resourceStore.delete(ex.Foo)
+
+      // when
+      await resourceStore.save()
+
+      // then
+      expect(client.query.update).to.have.been.calledOnce
```
Comment (lucafurrer) [inline apis/core/test/ResourceStore.test.ts:194]: @tpluscode what would be the good way to check that the graph does not exist anymore?

File: apis/core/hydra/csv-source.ttl [hunk 4]
```diff
@@ -57,6 +57,6 @@ cc:CSVSource
           a
             code:EcmaScript ;
           code:link
-            <file:handlers/csv-source#delete> ;
+            <file:handlers/csv-source#remove> ;
```
Comment (lucafurrer) [inline apis/core/hydra/csv-source.ttl:60]: Somehow the compiler was not too happy with a function called delete in the handler...

File: e2e-tests/csv-source/delete.hydra [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    Expect Operation cc:DeleteCSVSourceAction {
+        Invoke {
+        } => {
+            Expect Status 204
+        }     
+    }
+    Expect Operation cc:GetCSVSourceAction {
+        Invoke {
+        } => {
+            Expect Status 404
```
Comment (lucafurrer) [inline e2e-tests/csv-source/delete.hydra:70]: I would expect a 404 but a 405 is returned. Not sure if we should adapt the test or if there is a bug somewhere

**Review by tpluscode — CHANGES_REQUESTED** [review]

File: apis/core/test/support/TestResourceStore.ts [hunk 6]
```diff
@@ -3,13 +3,15 @@ import $rdf from 'rdf-ext'
...
+import TermSet from '@rdfjs/term-set'
 
 export class TestResourceStore implements ResourceStore {
   private readonly resources: TermMap<NamedNode, GraphPointer<NamedNode>>
+  private readonly __deletedGraphs: TermSet
 
   constructor(pointers: GraphPointer<NamedNode>[]) {
     this.resources = new TermMap()
-
+    this.__deletedGraphs = new TermSet()
```
[the remark this answers is not in the document]
Comment (lucafurrer) [reply apis/core/test/support/TestResourceStore.ts:14]: It is private...

[the remark this answers is not in the document]
Comment (tpluscode) [reply apis/core/test/support/TestResourceStore.ts:14]: The underscores is a convention which we don't enforce, but maybe we should. Ok, forget my suggestion for now

File: e2e-tests/csv-source/delete.hydra (hunk 5 above)
Comment (lucafurrer) [reply e2e-tests/csv-source/delete.hydra:70]: From a zulip discussion with @tpluscode 
`hydra-box looks up resources+operations by checking a resource named X (that is obviously getting removed) but then it also looks up resources used as objects by other resources. in our case it is a leftover <csv-mapping> cc:csvSource <deleted-source>`

File: apis/core/lib/domain/csv-source/delete.ts (hunk 2 above)
Comment (lucafurrer) [reply apis/core/lib/domain/csv-source/delete.ts:20]: Done -> #97

**Review by martinmaillard — APPROVED** [review]

**Review by tpluscode — APPROVED** [review]

**Comment by tpluscode** [comment]
Will also need a rebase or merge master

**Comment by lucafurrer** [comment]
I did it 1 hour ago, but here we go again

**Pull request merged**